test/unit/test-cookies.js

```diff
@@ -14,142 +14,175 @@
-    test('example.net', 'example.com', true);
-    test('cdn.ampproject.org', 'ampproject.org', true, true);
-    expect(() => {
-      test('cdn.ampproject.org', 'ampproject.org', true);
-    }).to.throw(/Should never attempt to set cookie on proxy origin\: c\&1/);
-    expect(() => {
-      test('CDN.ampproject.org', 'ampproject.org', true);
-    }).to.throw(/Should never attempt to set cookie on proxy origin\: c\&1/);
+
+    //example.com
+    test('https://www.example.com/test.html', 'example.com');
+    test('https://123.www.example.com/test.html', 'example.com');
+    test('https://example.com/test.html', 'example.com');
+    test('https://www.example.net', 'example.net');
+    doc.publicSufficList = ['example.com'];
+    test('https://123.www.example.com/test.html', 'www.example.com');
+  });
+
+  it('write cookie to right domain on proxy', () => {
+    win.location.href = 'https://foo.cdn.ampproject.org/test.html';
+    setCookie(win, 'c&1', 'v&1', Date.now() + BASE_CID_MAX_AGE_MILLIS, {
+      domain: 'foo.cdn.ampproject.org',
+      allowOnProxyOrigin: true,
+    });
+    expect(doc.lastSetCookieRaw).to.equal(
+      `c%261=v%261; path=/; domain=foo.cdn.ampproject.org; ` +
+        'expires=Tue, 01 Jan 2019 08:00:00 GMT'
+    );
+    expect(doc.cookie).to.equal('c%261=v%261');
+
+    // Fail if allowOnProxyOrigin is false
     expect(() => {
-      test('CDN.ampproject.org', 'AMPproject.org', true);
+      setCookie(win, 'c&1', 'v&1', Date.now() + BASE_CID_MAX_AGE_MILLIS, {});
     }).to.throw(/Should never attempt to set cookie on proxy origin\: c\&1/);
-    test('www.ampproject.org', 'www.ampproject.org');
-    test('cdn.ampproject.org', 'cdn.ampproject.org', false, true);
+
+    win.location.href = 'https://CDN.ampproject.org/test.html';
     expect(() => {
-      test('cdn.ampproject.org', 'cdn.ampproject.org', false);
+      setCookie(win, 'c&1', 'v&1', Date.now() + BASE_CID_MAX_AGE_MILLIS, {});
     }).to.throw(/Should never attempt to set cookie on proxy origin\: c\&1/);
+
+    win.location.href = 'https://foo.bar.cdn.ampproject.org/test.html';
     expect(() => {
-      test('foo.bar.cdn.ampproject.org', 'foo.bar.cdn.ampproject.org', false);
+      setCookie(win, 'c&1', 'v&1', Date.now() + BASE_CID_MAX_AGE_MILLIS, {});
     }).to.throw(/in depth check/);
+
+    win.location.href = 'http://&&&.CDN.ampproject.org/test.html';
     expect(() => {
-      test('&&&.cdn.ampproject.org', '&&&.cdn.ampproject.org', false);
+      setCookie(win, 'c&1', 'v&1', Date.now() + BASE_CID_MAX_AGE_MILLIS, {});
     }).to.throw(/in depth check/);
+
+    // Must have domain specified when allowOnProxyOrigin
     expect(() => {
-      test('&&&.CDN.ampproject.org', '&&&.cdn.AMPproject.org', false);
-    }).to.throw(/in depth check/);
+      setCookie(win, 'c&1', 'v&1', Date.now() + BASE_CID_MAX_AGE_MILLIS, {
+        allowOnProxyOrigin: true,
+      });
+    }).to.throw(/Should specify domain explicitly/);
+
+    // Cannot write to 'ampproject.org'
```

Comment:
Should we forbid writing to `cdn.ampproject.org` as well? I don't see a use case where we try to set cookie to `cdn.ampproject.org` using JavaScript.

Reply:
let's have a test case for that too

Reply:
You mean we want to change the code, forbid writing cookie to `cdn.ampproject.org` and then create a test case?

Reply:
done

Reply:
ah, wait, we do set cookies to cdn.ampproject.org in tools/experiments.js, 
and to xxx.cdn.ampproject.org in src/experiments.js

Reply:
ohhhh! That's a really good catch. Removed